File: backend/scripts/fix_unused_vars.py
#!/usr/bin/env python3
"""
Script pour corriger automatiquement les variables non utilisées
en les préfixant avec _ selon la convention ESLint
"""

import re
import subprocess
import sys
from pathlib import Path
from collections import defaultdict

def run_lint():
    """Exécute npm run lint et capture la sortie"""
    try:
        result = subprocess.run(
            ['npm', 'run', 'lint'],
            capture_output=True,
            text=True,
            cwd=Path(__file__).parent.parent
        )
        return result.stdout + result.stderr
    except Exception as e:
        print(f"❌ Erreur lors de l'exécution du lint: {e}")
        sys.exit(1)

def parse_warnings(lint_output):
    """Parse la sortie du lint pour extraire les warnings"""
    warnings = defaultdict(list)
    current_file = None
    
    for line in lint_output.split('\n'):
        # Détecte le fichier
        if line.startswith('/'):
            current_file = line.strip()
            continue
        
        # Détecte les warnings de variables non utilisées
        match = re.match(r'^\s+(\d+):(\d+)\s+warning\s+\'([^\']+)\' is (assigned a value but never used|defined but never used)', line)
        
        if match and current_file:
            line_num, col, var_name, _ = match.groups()
            
            # Ignorer si déjà préfixé avec _
            if not var_name.startswith('_'):
                warnings[current_file].append({
                    'line': int(line_num),
                    'col': int(col),
                    'var': var_name
                })
    
    return warnings

def fix_file(filepath, file_warnings):
    """Corrige un fichier donné"""
    try:
        with open(filepath, 'r', encoding='utf-8') as f:
            content = f.read()
            lines = content.split('\n')
        
        modified = False
        
        # Trier par ligne décroissante pour éviter les décalages
        for warning in sorted(file_warnings, key=lambda w: w['line'], reverse=True):
            line_idx = warning['line'] - 1
            if line_idx < 0 or line_idx >= len(lines):
                continue
            
            line = lines[line_idx]
            var_name = warning['var']
            
            # catch (error) → catch (_error)
            if f'catch ({var_name})' in line or f'catch({var_name})' in line:
                line = re.sub(rf'\bcatch\s*\(\s*{re.escape(var_name)}\s*\)', f'catch (_{{var_name}})', line)
                lines[line_idx] = line
                modified = True
                continue
            
            # const varName = → const _varName =
            if re.search(rf'\bconst\s+{re.escape(var_name)}\s*[=:]', line):
                line = re.sub(rf'\bconst\s+{re.escape(var_name)}\b', f'const _{var_name}', line)
                lines[line_idx] = line
                modified = True
                continue
            
            # let varName = → let _varName =
            if re.search(rf'\blet\s+{re_name}\s*=', line):
                line = re.sub(rf'\blet\s+{re.escape(var_name)}\b', f'let _{var_name}', line)
                lines[line_idx] = line
                modified = True
                continue
            
            # { data: varName, ... } → { data: _varName, ... }
            if re.search(rf':\s*{re.escape(var_name)}\s*[,}}]', line):
                line = re.sub(rf'(:\s*){re.escape(var_name)}(\s*[,}}])', rf'\1_{var_name}\2', line)
                lines[line_idx] = line
                modified = True
                continue
            
            # Import type { VarName } → Import type { VarName as _VarName }
            if 'import' in line.lower() and var_name in line:
                # Import simple: import { VarName } → import { type VarName as _VarName }
                if re.search(rf'\{[^}}]*\b{re.escape(var_name)}\b', line):
                    # Vérifier si c'est déjà un type import
                    if not re.search(rf'type\s+{re.escape(var_name)}\s+as\s+_', line):
                        line = re.sub(
                            rf'(\{{[^}}]*?)\b{re.escape(var_name)}\b([,\s]*[^}}]*\}})',
                            rf'\1type {var_name} as _{var_name}\2',
                            line
                        )
                        lines[line_idx] = line
                        modified = True
                        continue
            
            # Paramètres de fonction: functionName(varName: Type) → functionName(_varName: Type)
            if re.search(rf'\b{re.escape(var_name)}\s*:', line):
                # S'assurer que c'est bien un paramètre (dans une signature de fonction)
                if '(' in line or 'function' in line or '=>' in line:
                    line = re.sub(rf'\b{re.escape(var_name)}(\s*:)', rf'_{var_name}\1', line)
                    lines[line_idx] = line
                    modified = True
                    continue
        
        if modified:
            new_content = '\n'.join(lines)
            with open(filepath, 'w', encoding='utf-8') as f:
                f.write(new_content)
            return len(file_warnings)
        
        return 0
        
    except Exception as e:
        print(f"❌ Erreur avec {filepath}: {e}")
        return 0

def main():
    print("🔧 Correction automatique des variables non utilisées...\n")
    
    # Exécuter le lint
    print("📊 Analyse des warnings...")
    lint_output = run_lint()
    
    # Parser les warnings
    warnings = parse_warnings(lint_output)
    
    total_warnings = sum(len(w) for w in warnings.values())
    print(f"📊 Trouvé {total_warnings} warnings dans {len(warnings)} fichiers\n")
    
    if total_warnings == 0:
        print("✅ Aucun warning à corriger!")
        return
    
    # Corriger chaque fichier
    total_fixed = 0
    for filepath, file_warnings in warnings.items():
        if Path(filepath).exists():
            fixed = fix_file(filepath, file_warnings)
            if fixed > 0:
                print(f"✅ {filepath}: {fixed} corrections")
                total_fixed += fixed
    
    print(f"\n📊 Résumé: {total_fixed} warnings corrigés")
    print("🔍 Relancez 'npm run lint' pour vérifier\n")

if __name__ == '__main__':
    main()

File: backend/scripts/fix-unused-vars.js
#!/usr/bin/env node

/**
 * Script pour corriger automatiquement les variables non utilisées
 * en les préfixant avec _ selon la convention ESLint
 */

const fs = require('fs');
const { execSync } = require('child_process');

console.log('🔧 Correction automatique des variables non utilisées...\n');

// Exécuter le lint et capturer la sortie
let lintOutput;
try {
  execSync('npm run lint 2>&1', { encoding: 'utf-8', stdio: 'pipe' });
} catch (error) {
  lintOutput = error.stdout || error.stderr || '';
}

// Parser les warnings
const warnings = [];
const lines = lintOutput.split('\n');
let currentFile = null;

for (let i = 0; i < lines.length; i++) {
  const line = lines[i];
  
  // Détecter le fichier
  if (line.startsWith('/')) {
    currentFile = line.trim();
    continue;
  }
  
  // Détecter les warnings de variables non utilisées
  const warningMatch = line.match(/^\s+(\d+):(\d+)\s+warning\s+'([^']+)' is (assigned a value but never used|defined but never used)/);
  
  if (warningMatch && currentFile) {
    const [_, lineNum, col, varName, type] = warningMatch;
    
    // Ignorer les variables déjà préfixées avec _
    if (!varName.startsWith('_')) {
      warnings.push({
        file: currentFile,
        line: parseInt(lineNum),
        col: parseInt(col),
        varName,
        type
      });
    }
  }
}

console.log(`📊 Trouvé ${warnings.length} warnings à corriger\n`);

// Grouper par fichier
const fileWarnings = warnings.reduce((acc, w) => {
  if (!acc[w.file]) acc[w.file] = [];
  acc[w.file].push(w);
  return {};
}, {});

let fixed = 0;
let errors = 0;

// Traiter chaque fichier
Object.entries(fileWarnings).forEach(([file, fileWarnings]) => {
  try {
    if (!fs.existsSync(file)) {
      console.log(`⚠️  Fichier non trouvé: ${file}`);
      return;
    }
    
    let content = fs.readFileSync(file, 'utf-8');
    const lines = content.split('\n');
    
    // Trier par numéro de ligne décroissant pour ne pas décaler les lignes
    fileWarnings.sort((a, b) => b.line - a.line);
    
    fileWarnings.forEach(warning => {
      const lineIndex = warning.line - 1;
      if (lineIndex < 0 || lineIndex >= lines.length) return;
      
      const line = lines[lineIndex];
      const varName = warning.varName;
      
      // Patterns de remplacement sûrs
      let newLine = line;
      let modified = false;
      
      // catch (error) → catch (_error)
      if (line.includes(`catch (${varName})`)) {
        newLine = line.replace(`catch (${varName})`, `catch (_${varName})`);
        modified = true;
      }
      // const varName = → const _varName =
      else if (line.match(new RegExp(`\\bconst\\s+${varName}\\s*[=:]`))) {
        newLine = line.replace(new RegExp(`\\bconst\\s+${varName}\\b`), `const _${varName}`);
        modified = true;
      }
      // let varName = → let _varName =
      else if (line.match(new RegExp(`\\blet\\s+${varName}\\s*=`))) {
        newLine = line.replace(new RegExp(`\\blet\\s+${varName}\\b`), `let _${varName}`);
        modified = true;
      }
      // Destructuring: { varName, ... }
      else if (line.match(new RegExp(`\\{[^}]*\\b${varName}\\b[^}]*\\}`))) {
        // Plus complexe, on préfère le skip pour éviter les bugs
        console.log(`⚠️  Skip destructuring: ${file}:${warning.line} (${varName})`);
      }
      // Import { VarName }
      else if (line.match(new RegExp(`import\\s*\\{[^}]*\\b${varName}\\b`))) {
        newLine = line.replace(
          new RegExp(`(\\b${varName}\\b)([,}])`),
          `type ${varName} as _${varName}$2`
        );
        modified = true;
      }
      
      if (modified) {
        lines[lineIndex] = newLine;
        fixed++;
      }
    });
    
    // Sauvegarder si modifié
    const newContent = lines.join('\n');
    if (newContent !== content) {
      fs.writeFileSync(file, newContent, 'utf-8');
      console.log(`✅ ${file} (${fileWarnings.length} corrections)`);
    }
    
  } catch (error) {
    console.log(`❌ Erreur avec ${file}: ${error.message}`);
    errors++;
  }
});

console.log(`\n📊 Résumé:`);
console.log(`   ✅ ${fixed} warnings corrigés`);
console.log(`   ❌ ${errors} erreurs`);
console.log(`\n🔍 Relancez 'npm run lint' pour vérifier`);
